### PR TITLE
Add rust typechecker of tree IR

### PR DESCRIPTION
**Pull request by oflatt**
This PR eases the burden of translation from RVSDG to tree encoding by automatically annotating the types of every argument reference in the program.

### CODE REVIEW

File: tree_assume/src/ast.rs [hunk 1]
```diff
@@ -201,12 +203,20 @@ pub fn tlet(lhs: RcExpr, rhs: RcExpr) -> RcExpr {
     RcExpr::new(Expr::Let(lhs, rhs))
 }
 
-pub fn arg(ty: Type) -> RcExpr {
-    RcExpr::new(Expr::Arg(ty))
+/// Assumes the argument has an integer type.
+/// If this is not true, use `with_arg_types` to add
+/// the correct types to arguments after construction.
+pub fn arg() -> RcExpr {
```
Comment (ajpal) [inline tree_assume/src/ast.rs:209]: call this `int_arg`?

File: tree_assume/src/optimizations/is_valid.rs [hunk 2]
```diff
@@ -40,11 +40,12 @@ fn test_is_valid() -> Result<(), egglog::Error> {
...
         parallel!(
-            less_than(get(arg(intt()), 0), int(3)),
+            less_than(get(arg(), 0), int(3)),
             get(switch!(int(0); parallel!(int(4), int(5))), 0)
         ),
-    );
-    let not_made_valid = sub(arg(intt()), arg(intt())); // this expression is indeed valid, but won't be marked as so
+    )
+    .with_arg_types(emptyt(), tuplet!(intt()));
+    let not_made_valid = sub(arg(), arg()).with_arg_types(intt(), intt()); // this expression is indeed valid, but won't be marked as so
```
Comment (ajpal) [inline tree_assume/src/optimizations/is_valid.rs:48]: what does this comment mean? (ik not added by this diff)

File: tree_assume/src/typechecker.rs [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+                .iter()
+                .map(|expr| self.add_arg_types_to_func(expr.clone()))
+                .collect(),
+        }
+    }
+
+    pub(crate) fn add_arg_types_to_func(&self, func: RcExpr) -> RcExpr {
+        match func.as_ref() {
+            Expr::Function(name, in_ty, out_ty, body) => {
+                let (ety, new_body) = self.add_arg_types_to_expr(body.clone(), in_ty.clone());
```
Comment (ajpal) [inline tree_assume/src/typechecker.rs:62]: `ety` is a confusing name imo. `expr_ty`?

File: tree_assume/src/typechecker.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+                };
+                assert_eq!(out_tys[0], Type::Base(BaseType::BoolT));
+                assert_eq!(out_tys[1..], in_tys);
+                (
+                    Type::TupleT(out_tys[1..].to_vec()),
+                    RcExpr::new(Expr::DoWhile(new_inputs, new_pred_and_outputs)),
+                )
+            }
+            // Replace the argument type with the new type
+            Expr::Arg(_ignored_ty) => (arg_ty.clone(), Rc::new(Expr::Arg(arg_ty))),
```
Comment (ajpal) [inline tree_assume/src/typechecker.rs:290]: why do we ignore the type?

File: tree_assume/src/optimizations/is_valid.rs (hunk 2 above)
Comment (oflatt) [reply tree_assume/src/optimizations/is_valid.rs:48]: clarified-
```
// this expression is valid (it uses only IR constructors)
    // but it isn't a sub-expression of the initial one, so it won't be
    // marked as valid.```

File: tree_assume/src/typechecker.rs (hunk 4 above)
Comment (oflatt) [reply tree_assume/src/typechecker.rs]: no longer ignored- does type checking, and replaces unknowns

File: tree_assume/src/ast.rs [hunk 5]
```diff
@@ -201,12 +203,24 @@ pub fn tlet(lhs: RcExpr, rhs: RcExpr) -> RcExpr {
...
+    RcExpr::new(Expr::Arg(intt()))
+}
+
+/// An argument with a boolean type.
+pub fn bool_arg() -> RcExpr {
+    RcExpr::new(Expr::Arg(boolt()))
+}
+
+pub fn getat(index: usize) -> RcExpr {
+    get(int_arg(), index)
```
Comment (ajpal) [inline tree_assume/src/ast.rs:223]: should this be unknown arg?

File: tree_assume/src/schema.rs [hunk 6]
```diff
@@ -16,6 +16,12 @@ pub enum Type {
     PointerT(BaseType),
     /// Nested tuple types are not allowed.
     TupleT(Vec<Type>),
+    /// Before `with_arg_types`, users of this IR can leave unknown types
+    /// in arguments.
+    /// When all types are present except for unknowns in arguments,
+    /// `with_arg_types` succeeds and the unknowns are replaced with the correct types.
+    /// In the egraph there should never be any unknown arg types.
```
Comment (ajpal) [inline tree_assume/src/schema.rs:23]: add comment: this is enforced by `to_egglog` so readers here don't worry

**Review by ajpal — APPROVED** [review]

File: tree_assume/src/ast.rs (hunk 5 above)
Comment (oflatt) [reply tree_assume/src/ast.rs:223]: good catch!

**Pull request merged**